### PR TITLE
Python: Add astra memory

### PR DESCRIPTION
**Pull request by david-polo**
### Motivation and Context

We are going to add [Astra](https://docs.datastax.com/en/home/docs/index.html) as a memory store.

DataStax Astra DB is a serverless vector database that’s perfect for managing mission-critical AI workloads. It’s built on Apache Cassandra®, making Astra DB a highly scalable, reliable database technology. It’s a powerful all-in-one data storage solution, ideal for Generative AI projects.

### Description

- Add astra into python/semantic_kernel/connectors/memory/astradb
- Add test into python/tests/integration/connectors/memory/test_astradb.py

### Contribution Checklist

- [ ] The code builds clean without any errors or warnings
- [ ] The PR follows the [SK Contribution Guidelines](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md) and the [pre-submission formatting script](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md#development-scripts) raises no violations
- [ ] All unit tests pass, and I have added new tests where possible
- [ ] I didn't break anyone :smile:

### CODE REVIEW

File: python/tests/integration/connectors/memory/test_astradb.py
Comment (hc33brackles) [reply python/tests/integration/connectors/memory/test_astradb.py]: Hi @eavanvalkenburg  - our team updated the code so we are now passing all tests with the functions converted to Async

File: python/semantic_kernel/connectors/memory/astradb/astradb_memory_store.py
Comment (hc33brackles) [reply python/semantic_kernel/connectors/memory/astradb/astradb_memory_store.py:68]: Hi @eavanvalkenburg - our team updated the code so we are now passing all tests with the functions converted to Async

File: python/semantic_kernel/connectors/memory/astradb/astra_client.py
Comment (hc33brackles) [reply python/semantic_kernel/connectors/memory/astradb/astra_client.py]: done

File: python/tests/integration/connectors/memory/test_astradb.py
Comment (hc33brackles) [reply python/tests/integration/connectors/memory/test_astradb.py]: done

File: python/semantic_kernel/connectors/memory/astradb/astradb_memory_store.py
Comment (hc33brackles) [reply python/semantic_kernel/connectors/memory/astradb/astradb_memory_store.py]: done

File: python/semantic_kernel/connectors/memory/astradb/astradb_memory_store.py
Comment (hc33brackles) [reply python/semantic_kernel/connectors/memory/astradb/astradb_memory_store.py]: done

**Comment by hc33brackles** [comment]
@eavanvalkenburg I've reviewed the changes you requested with my team and we are good to go

File: python/semantic_kernel/connectors/memory/astradb/astradb_memory_store.py
Comment (eavanvalkenburg) [reply python/semantic_kernel/connectors/memory/astradb/astradb_memory_store.py:68]: Thanks, but the methods in the client themselves should be async, this way it will run new loops many times leading to all sorts of weird behaviours, so it is going async main -> sync here -> async.run(async method in client) which is really not necessary and not the way it should be.

File: python/semantic_kernel/connectors/memory/astradb/astra_client.py
Comment (eavanvalkenburg) [inline python/semantic_kernel/connectors/memory/astradb/astra_client.py:142]: do not use asyncio.run inside here, just make the method async and do `await self._run_query`

File: python/semantic_kernel/connectors/memory/astradb/astra_client.py
Comment (eavanvalkenburg) [inline python/semantic_kernel/connectors/memory/astradb/astra_client.py:34]: I would like to see the option to supply your own session for this, have a look at this PR for how to do that: #4477

File: python/semantic_kernel/connectors/memory/astradb/exceptions.py
Comment (eavanvalkenburg) [inline python/semantic_kernel/connectors/memory/astradb/exceptions.py:1]: Can this be based on the existing exception classes within SK?

**Review by eavanvalkenburg — CHANGES_REQUESTED** [review]

File: python/semantic_kernel/connectors/memory/astradb/astra_client.py
Comment (eavanvalkenburg) [inline python/semantic_kernel/connectors/memory/astradb/astra_client.py:14]: the option to add the session to the init is missing.

**Review by eavanvalkenburg — CHANGES_REQUESTED** [review]
My main concern of the previous review is not yet addressed. The methods inside the astra_client should be async and never call asyncio.run, nest_asyncio should not be needed!

File: python/semantic_kernel/connectors/memory/astradb/astradb_memory_store.py
Comment (hc33brackles) [reply python/semantic_kernel/connectors/memory/astradb/astradb_memory_store.py:73]: Okay- this is now done too @eavanvalkenburg - thanks for the clarification

File: python/semantic_kernel/connectors/memory/astradb/astra_client.py
Comment (hc33brackles) [reply python/semantic_kernel/connectors/memory/astradb/astra_client.py]: Done

File: python/semantic_kernel/connectors/memory/astradb/astra_client.py
Comment (hc33brackles) [reply python/semantic_kernel/connectors/memory/astradb/astra_client.py]: done

**Comment by hc33brackles** [comment]
@eavanvalkenburg  - it looks to me like we have finalized all the changes requested

File: python/semantic_kernel/connectors/memory/astradb/astra_client.py
Comment (hc33brackles) [reply python/semantic_kernel/connectors/memory/astradb/astra_client.py:11]: done

File: python/semantic_kernel/connectors/memory/astradb/exceptions.py
Comment (hc33brackles) [reply python/semantic_kernel/connectors/memory/astradb/exceptions.py]: done

**Comment by hc33brackles** [comment]
@eavanvalkenburg 

Please let me know if there is anything else we can do to support you and get this merged. If we are missing something we needed to do can you please point it out?

**Comment by hc33brackles** [comment]
Bumping this @eavanvalkenburg - do we need to do anything else? cc: @david-polo

File: python/semantic_kernel/connectors/memory/astradb/astradb_memory_store.py
Comment (eavanvalkenburg) [inline python/semantic_kernel/connectors/memory/astradb/astradb_memory_store.py:39]: we no longer use this setup for loggers, please remove and use regular "pythonic" logging (`import logging, logger = logging.get_logger(__name__)`)

File: python/semantic_kernel/connectors/memory/astradb/astradb_memory_store.py
Comment (eavanvalkenburg) [inline python/semantic_kernel/connectors/memory/astradb/astradb_memory_store.py:67]: do we need these non-async methods?

File: python/semantic_kernel/connectors/memory/astradb/astradb_memory_store.py
Comment (eavanvalkenburg) [inline python/semantic_kernel/connectors/memory/astradb/astradb_memory_store.py:73]: and this is calling a async method, so this will not work anyway

File: python/semantic_kernel/connectors/memory/astradb/astradb_memory_store.py
Comment (eavanvalkenburg) [inline python/semantic_kernel/connectors/memory/astradb/astradb_memory_store.py:166]: could this be parallelized by using asyncio.gather?

**Review by eavanvalkenburg — CHANGES_REQUESTED** [review]
some minor notes, starting to look good!

File: python/semantic_kernel/connectors/memory/astradb/astradb_memory_store.py
Comment (david-polo) [reply python/semantic_kernel/connectors/memory/astradb/astradb_memory_store.py]: Done

File: python/semantic_kernel/connectors/memory/astradb/astradb_memory_store.py
Comment (david-polo) [reply python/semantic_kernel/connectors/memory/astradb/astradb_memory_store.py]: Done

File: python/semantic_kernel/connectors/memory/astradb/astradb_memory_store.py
Comment (david-polo) [reply python/semantic_kernel/connectors/memory/astradb/astradb_memory_store.py]: Done

File: python/semantic_kernel/connectors/memory/astradb/astradb_memory_store.py
Comment (david-polo) [reply python/semantic_kernel/connectors/memory/astradb/astradb_memory_store.py]: Done

File: python/semantic_kernel/connectors/memory/astradb/astradb_memory_store.py
Comment (david-polo) [reply python/semantic_kernel/connectors/memory/astradb/astradb_memory_store.py]: Done

**Comment by hc33brackles** [comment]
Hi @eavanvalkenburg  - all of these comments are resolved.

File: python/semantic_kernel/connectors/memory/astradb/astra_client.py
Comment (eavanvalkenburg) [inline python/semantic_kernel/connectors/memory/astradb/astra_client.py:48]: nit: do we need the brackets around the call?

File: python/semantic_kernel/connectors/memory/astradb/astra_client.py
Comment (eavanvalkenburg) [inline python/semantic_kernel/connectors/memory/astradb/astra_client.py:70]: same

File: python/semantic_kernel/connectors/memory/astradb/astra_client.py
Comment (eavanvalkenburg) [inline python/semantic_kernel/connectors/memory/astradb/astra_client.py:78]: this returns the collection_url not the query, please rename!

File: python/semantic_kernel/connectors/memory/astradb/astradb_memory_store.py
Comment (eavanvalkenburg) [inline python/semantic_kernel/connectors/memory/astradb/astradb_memory_store.py:29]: please move this outside of the class, below the MAX_DELETE_BATCH_SIZE line, we want logging to be handled uniformly, then any logging calls just just `logger.` instead of the current `self.logger.`

File: python/semantic_kernel/connectors/memory/astradb/astradb_memory_store.py
Comment (eavanvalkenburg) [inline python/semantic_kernel/connectors/memory/astradb/astradb_memory_store.py:122]: please do this in one line by using `logger.log(logging.INFO if result else logging.WARNING, f"Collection {....`.

File: python/semantic_kernel/connectors/memory/astradb/astradb_memory_store.py
Comment (eavanvalkenburg) [inline python/semantic_kernel/connectors/memory/astradb/astradb_memory_store.py:160]: please add a docstring here, explaining limits and approaches to batching!

File: python/semantic_kernel/connectors/memory/astradb/utils.py
Comment (eavanvalkenburg) [inline python/semantic_kernel/connectors/memory/astradb/utils.py:35]: Pinecone?

File: python/semantic_kernel/connectors/memory/astradb/utils.py
Comment (eavanvalkenburg) [inline python/semantic_kernel/connectors/memory/astradb/utils.py:19]: Pinecone?

File: python/semantic_kernel/connectors/memory/astradb/utils.py
Comment (eavanvalkenburg) [inline python/semantic_kernel/connectors/memory/astradb/utils.py:37]: why not just use `document.get('text')`?

**Review by eavanvalkenburg — CHANGES_REQUESTED** [review]
mostly some minor things todo, thanks again for the efforts!

File: python/semantic_kernel/connectors/memory/astradb/astra_client.py
Comment (david-polo) [reply python/semantic_kernel/connectors/memory/astradb/astra_client.py:48]: Remove the brackets.

File: python/semantic_kernel/connectors/memory/astradb/astra_client.py
Comment (david-polo) [reply python/semantic_kernel/connectors/memory/astradb/astra_client.py:70]: Remove the brackets.

File: python/semantic_kernel/connectors/memory/astradb/astra_client.py
Comment (david-polo) [reply python/semantic_kernel/connectors/memory/astradb/astra_client.py]: Done

File: python/semantic_kernel/connectors/memory/astradb/astradb_memory_store.py
Comment (david-polo) [reply python/semantic_kernel/connectors/memory/astradb/astradb_memory_store.py]: Done

File: python/semantic_kernel/connectors/memory/astradb/astradb_memory_store.py
Comment (david-polo) [reply python/semantic_kernel/connectors/memory/astradb/astradb_memory_store.py]: Done

File: python/semantic_kernel/connectors/memory/astradb/astradb_memory_store.py
Comment (david-polo) [reply python/semantic_kernel/connectors/memory/astradb/astradb_memory_store.py:160]: Done

File: python/semantic_kernel/connectors/memory/astradb/utils.py
Comment (david-polo) [reply python/semantic_kernel/connectors/memory/astradb/utils.py]: Removed

File: python/semantic_kernel/connectors/memory/astradb/utils.py
Comment (david-polo) [reply python/semantic_kernel/connectors/memory/astradb/utils.py]: Removed

**Comment by david-polo** [comment]
Hi @eavanvalkenburg - all of these comments are resolved.

File: python/tests/integration/connectors/memory/test_astradb.py
Comment (moonbox3) [inline python/tests/integration/connectors/memory/test_astradb.py:108]: Do we need these xfail decorators that are commented out?

**Review by moonbox3 — COMMENTED** [review]
Looking fine to me, overall. Just one small question about some commented out code -- if it's necessary.

**Comment by moonbox3** [comment]
Please also run `black` formatting on your code as there are some failures.

File: python/semantic_kernel/connectors/memory/astradb/astradb_memory_store.py
Comment (eavanvalkenburg) [inline python/semantic_kernel/connectors/memory/astradb/astradb_memory_store.py:167]: This should be List[MemoryRecord] right?

File: python/semantic_kernel/connectors/memory/astradb/astradb_memory_store.py
Comment (eavanvalkenburg) [inline python/semantic_kernel/connectors/memory/astradb/astradb_memory_store.py:172]: why does this method not user update_documents? seems like there is support for native batching on the API side which should be much faster?

**Review by eavanvalkenburg — CHANGES_REQUESTED** [review]
two small notes left!

File: python/semantic_kernel/connectors/memory/astradb/astradb_memory_store.py
Comment (david-polo) [reply python/semantic_kernel/connectors/memory/astradb/astradb_memory_store.py:172]: There's an astra API for batching. But it can't update by the list of id.
`curl -s --location \
--request POST ${ASTRA_DB_API_ENDPOINT}/api/json/v1/${ASTRA_DB_KEYSPACE}/${COLLECTION} \
--header "Token: ${ASTRA_DB_APPLICATION_TOKEN}" \
--header "Content-Type: application/json" \
--header "Accept: application/json" \
--data '{
  "updateMany": {
    "filter": {"status" : "active" },
    "update" : {"$set" : { "status" : "inactive"}}
  }
}' | json_pp`

As you can see, can't pass multiple ids to filter option.

File: python/semantic_kernel/connectors/memory/astradb/astradb_memory_store.py
Comment (david-polo) [reply python/semantic_kernel/connectors/memory/astradb/astradb_memory_store.py:167]: Fxied

File: python/semantic_kernel/connectors/memory/astradb/astradb_memory_store.py
Comment (david-polo) [reply python/semantic_kernel/connectors/memory/astradb/astradb_memory_store.py:172]: @eavanvalkenburg

Comment (hc33brackles) [reply python/semantic_kernel/connectors/memory/astradb/astradb_memory_store.py:172]: done

**Comment by hc33brackles** [comment]
@eavanvalkenburg  - we are done with those two comments and ready for whatever is the next step 👍

File: python/tests/integration/connectors/memory/test_astradb.py
Comment (moonbox3) [reply python/tests/integration/connectors/memory/test_astradb.py:108]: @hc33brackles quick ping on this

**Comment by moonbox3** [comment]
> Please also run `black` formatting on your code as there are some failures.

@david-polo please run `black .` in the `semantic-kernel/python` directory to make sure all formatting is correct. You also need to run a `ruff . --fix` in that same directory to fix ruff formatting issues.

**Comment by david-polo** [comment]
@moonbox3 , I fixed lint issue.

**Comment by moonbox3** [comment]
> @moonbox3 , I fixed lint issue.

@david-polo the `black` formatting check is still failing. Please address.

**Comment by david-polo** [comment]
@moonbox3 , Done.
All passed.

**Review by moonbox3 — APPROVED** [review]
LGTM. We'll need @eavanvalkenburg to give an approval as well, since he has requested changes (and we require 2 approvers).

File: python/semantic_kernel/connectors/memory/astradb/utils.py
Comment (eavanvalkenburg) [reply python/semantic_kernel/connectors/memory/astradb/utils.py:21]: Probably just `Dict[str, Any]`

Comment (david-polo) [reply python/semantic_kernel/connectors/memory/astradb/utils.py]: Done

Comment (david-polo) [reply python/semantic_kernel/connectors/memory/astradb/utils.py]: @eavanvalkenburg Fixed!

**Comment by hc33brackles** [comment]
@eavanvalkenburg  - we think this is all done and good to go -- please let us know if we are missing anything and there's more left to do. cc: @david-polo

**Review by eavanvalkenburg — APPROVED** [review]
Looks good, thanks for your efforts and patience!

File: python/tests/integration/connectors/memory/test_astradb.py
Comment (moonbox3) [inline python/tests/integration/connectors/memory/test_astradb.py:15]: I thought I remembered seeing this handled differently. We can't have `astradb_installed` default to True, and then have `pytest.mark.skipip(...)` with no other logic in between. We need the ability to set `astradb_installed` to false. Please follow a pattern similar to other connector integration tests, as is done [here](https://github.com/microsoft/semantic-kernel/blob/4317f2745e22055eee9eae811d7a4c48d557948d/python/tests/integration/connectors/memory/test_mongodb_atlas.py#L16).

**Review by moonbox3 — CHANGES_REQUESTED** [review]
Let's please fix the astradb initialization logic in the integration test.

File: python/tests/integration/connectors/memory/test_astradb.py
Comment (hc33brackles) [reply python/tests/integration/connectors/memory/test_astradb.py:15]: We are working on this now.

Comment (anomnaco) [reply python/tests/integration/connectors/memory/test_astradb.py:19]: Fixed: astradb_installed is set to false if the memory store or reqs fail to import

Comment (hc33brackles) [reply python/tests/integration/connectors/memory/test_astradb.py:19]: @moonbox3 @eavanvalkenburg there's a fix for this pushed

**Review by moonbox3 — APPROVED** [review]

File: python/tests/integration/connectors/memory/test_astradb.py
Comment (moonbox3) [reply python/tests/integration/connectors/memory/test_astradb.py:19]: @hc33brackles, @david-polo, @anomnaco, please run `black .` in the `semantic-kernel/python` directory and push the changes. Once in, I can try and merge to main. Thanks!

**Review by moonbox3 — APPROVED** [review]

**Comment by hc33brackles** [comment]
We noticed this got removed from the merge queue due to some failing tests, taking a look at this now: https://github.com/microsoft/semantic-kernel/actions/runs/7703289774/job/20993370549

File: python/tests/integration/connectors/memory/test_astradb.py
Comment (moonbox3) [inline python/tests/integration/connectors/memory/test_astradb.py:15]: @david-polo how would this env variable get populated? I don't see it present in the `.env.example` that was updated for this PR.

Is there a specific `astradb` package that needs to be added to python's pyproject.toml via a `poetry add <dependency>`? Is the dependency called `astrapy`? 

Couldn't we check if the `ASTRADB_APP_TOKEN` is present?

**Review by moonbox3 — APPROVED** [review]

**Comment by hc33brackles** [comment]
@eavanvalkenburg 

 We've been trying to use the SK integration recently and it's no longer working. We've been trying to run the same integration tests that passed at the time we merged the integration (the PR: https://github.com/microsoft/semantic-kernel/pull/4374). We're getting this error:

TypeError: Can't instantiate abstract class AstraDBMemoryStore with abstract methods create_collection, delete_collection, does_collection_exist, get, get_batch, get_collections, get_nearest_match, get_nearest_matches, remove, remove_batch, upsert, upsert_batch

Which is confusing because AstraDBMemoryStore isn't an abstract class. It appears that most integrations for memory store are not abstract classes. Can you (or anyone) help us understand this problem? We've also reported this issue here: https://github.com/microsoft/semantic-kernel/issues/5092

**Pull request merged**
